src/libmongoc/src/mongoc/mongoc-client-session-private.h

```diff
@@ -38,14 +38,9 @@ struct _mongoc_transaction_opt_t {
    int64_t max_commit_time_ms;
 };
 
-typedef enum {
-   MONGOC_SESSION_NO_OPTS = 0,
-   MONGOC_SESSION_CAUSAL_CONSISTENCY = (1 << 0),
-   MONGOC_SESSION_SNAPSHOT = (2 << 0),
-} mongoc_session_flag_t;
-
 struct _mongoc_session_opt_t {
-   mongoc_session_flag_t flags;
+   mongoc_optional_t causal_consistency;
+   mongoc_optional_t snapshot;
```

Comment:
Seems better to have these as optionals instead of on the same flag, thanks. Should the type be `mongoc_optional_bool_t` or is that type name change API-breaking?

Reply:
I brought that up with @kevinAlbs in Slack last week and we concluded that the ship has sailed since `mongoc_optional_t` is already in the public API (shipped with Versioned API in 1.18.0). If we end up introducing an optional int type down the line, perhaps we can deprecate the original name and introduce `mongoc_optional_bool_t` alongside `mongoc_optional_int_t`, but I don't think there's anything to do presently.

Reply:
Good catch @jmikola!

Here is my understanding.

Before snapshot reads:
- `mongoc_session_opts_new` sets `causalConsistency=true`.

After snapshot reads:
- `mongoc_session_opts_new` sets `causalConsistency=true` and `snapshot=false`.
- Validation occurs in `mongoc_client_start_session` (i.e. both being true is an error).

The new behavior proposed in this PR:
- `mongoc_session_opts_new` sets `causalConsistency=unset` and `snapshot=unset`
- Defaults are applied in `mongoc_client_start_session` if the values are unset. Validation occurs in `mongoc_client_start_session`.

Admittedly, when reviewing the snapshot reads changes, I had thought the behavior was correct since it built off of the previous behavior. I had not considered introducing an unset value for `causalConsistency` and `snapshot`. This seems correct.

Reply:
Your understanding is correct.

I think much of the confusion actually relates in some way to [SPEC-936](https://jira.mongodb.org/browse/SPEC-936), which added the technical equivalent of legalese to state that the `causalConsistency` session option doesn't get a default value (which would require modeling it as a trilean), but instead inherits some default. In practice, I think some drivers (libmongoc included) just defaulted it to true as that had a functionally equivalent effect.